framework/Optimizers/SimulatedAnnealing.py

```diff
@@ -0,0 +1,718 @@
+    fraction = iter/self.limit
+    currentPoint = self._collectOptPoint(rlz)
+    self.T0 = self._temperature(fraction)
+    self.T = self._coolingSchedule(iter,self.T0)
+    if traj in self._activeTraj:
+      newPoint = self._nextNeighbour(rlz,fraction)
+      # check new opt point against constraints
+      try:
+        suggested, modded = self._handleExplicitConstraints(newPoint, currentPoint, 'opt')
+      except NoConstraintResolutionFound:
+        # we've tried everything, but we just can't hack it
+        self.raiseAMessage('Optimizer "{}" trajectory {} was unable to continue due to functional or boundary constraints.'
+                          .format(self.name, traj))
+        self._closeTrajectory(traj, 'converge', 'no constraint resolution', newPoint[self._objectiveVar])
+        return
+      self._submitRun(suggested, traj, self.getIteration(traj))
+
+  # * * * * * * * * * * * * * * * *
+  # Convergence Checks
+  # Note these names need to be formatted according to checkConvergence check!
+  convFormat = ' ... {name:^12s}: {conv:5s}, {got:1.2e} / {req:1.2e}'
+
+  # NOTE checkConvSamePoint has a different call than the others
+  # should this become an informational dict that can be passed to any of them?
+
+  def checkConvergence(self, traj, new, old):
+    """
+      Check for trajectory convergence
+      @ In, traj, int, trajectory to consider
+      @ Out, None? FIXME
+    """
+    convs = {}
+    for conv in self._convergenceCriteria:
+      # special treatment for same point check
+      if conv == 'samePoint':
+        convs[conv] = self._checkConvSamePoint(new, old)
+        continue
+      # fix capitalization for RAVEN standards
+      fName = conv[:1].upper() + conv[1:]
+      # get function from lookup
+      f = getattr(self, '_checkConv{}'.format(fName))
+      # check convergence function
+      okay = f(traj)
+      # store and update
+      convs[conv] = okay
+    return any(convs.values()), convs
+
+  def _checkConvSamePoint(self, new, old):
+    """
+      Checks for a repeated same point
+      @ In, new, dict, new opt point
+      @ In, old, dict, old opt point
+      @ Out, converged, bool, convergence state
+    """
+    # TODO diff within tolerance? Exactly equivalent seems good for now
+    same = list(abs(new[var] - old[var])==self._convergenceCriteria['samePoint'] for var in self.toBeSampled)
+    converged = all(same)
+    self.raiseADebug(self.convFormat.format(name='same point',
+                                            conv=str(converged),
+                                            got=sum(same),
+                                            req=len(same)))
+    return converged
+  def _checkConvObjective(self, traj):
+    """
+      Checks the change in objective for convergence
+      @ In, traj, int, trajectory identifier
+      @ Out, converged, bool, convergence state
+    """
```

Comment:
is this copied/pasted? is it needed?

Reply:
where do you use it?

Reply:
yes

Reply:
yes? 
where do you use it?

Reply:
I can not find a place in your module where you use it... if it is not used, why do you need it here?

Reply:
As mentioned before:
`      # fix capitalization for RAVEN standards
      fName = conv[:1].upper() + conv[1:]
      # get function from lookup
      f = getattr(self, '_checkConv{}'.format(fName))
      # check convergence function
      okay = f(traj)
      # store and update
      convs[conv] = okay`
calls the functions _checkConvObjective, _checkConvTemperature